cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinCondition.java

```diff
@@ -0,0 +1,226 @@
+
+    @Override
+    public void validate(List<JoinStage> joinStages) {
+      super.validate(joinStages);
+
+      Map<String, JoinStage> stageMap = joinStages.stream()
+        .collect(Collectors.toMap(JoinStage::getStageName, s -> s));
+
+      for (JoinKey joinKey : keys) {
+        String joinStageName = joinKey.getStageName();
+        JoinStage joinStage = stageMap.get(joinStageName);
+        // check that the stage for each key is in the list of stages
+        if (joinStage == null) {
+          throw new InvalidJoinConditionException(String.format(
+            "Join key for stage '%s' is invalid. Stage '%s' is not a join input.",
+            joinStageName, joinStageName));
+        }
+        // this happens if the schema for that stage is unknown.
+        // for example, because of macros could not be evaluated yet.
+        if (joinStage.getSchema() == null) {
+          continue;
+        }
+        Set<String> fields = joinStage.getSchema().getFields().stream()
+          .map(Schema.Field::getName)
+          .collect(Collectors.toSet());
+
+        // check that the key fields for each key is in fields for that stage
+        // for example, when joining on A.id = B.uid, check that 'id' is in the fields for stage A and 'uid' for stage B
+        Set<String> keysCopy = new HashSet<>(joinKey.getFields());
+        keysCopy.removeAll(fields);
+        if (keysCopy.size() == 1) {
+          throw new InvalidJoinConditionException(String.format(
+            "Join key for stage '%s' is invalid. Field '%s' does not exist in the stage.",
+            joinStageName, keysCopy.iterator().next()));
+        }
+        if (keysCopy.size() > 1) {
+          throw new InvalidJoinConditionException(String.format(
+            "Join key for stage '%s' is invalid. Fields %s do not exist in the stage.",
+            joinStageName, String.join(", ", keysCopy)));
+        }
+      }
+
+      // check that the keys have the same type.
+
+      Iterator<JoinKey> keyIter = keys.iterator();
+      JoinKey key1 = keyIter.next();
+      // stage is guaranteed to be exist because of above validation
+      Schema schema1 = stageMap.get(key1.getStageName()).getSchema();
+      if (schema1 == null) {
+        return;
+      }
+      while (keyIter.hasNext()) {
+        JoinKey otherKey = keyIter.next();
+        Schema otherSchema = stageMap.get(otherKey.getStageName()).getSchema();
+        if (otherSchema == null) {
+          continue;
+        }
+        Iterator<String> fieldIter = otherKey.getFields().iterator();
+        for (String key1Field : key1.getFields()) {
+          Schema key1Schema = getNonNullableFieldSchema(schema1, key1Field);
+
+          // don't need to check hasNext() because already verified they are all the same length
+          String otherField = fieldIter.next();
+          Schema otherFieldSchema = getNonNullableFieldSchema(otherSchema, otherField);
+
+          if (key1Schema.getType() != otherFieldSchema.getType()) {
```

Comment:
should we also compare logical types?

Reply:
I thought about that and wavered back and forth. 

In the end, I decided to allow it because it is actually able to join if the physical types are the same, regardless of the logical types. It probably makes sense for the plugin to check this and disallow, but didn't want to restrict it in the app.

Reply:
I see. sg 👍